### PR TITLE
[#3337] improvement(hadoop-catalog): Support user impersonation for Hadoop catalog.

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

Add user impersonation for the Hadoop catalog.

### Why are the changes needed?

We need authentication for the encrypted HDFS cluster.

Fix: #3337

### Does this PR introduce _any_ user-facing change?

N/A. 

### How was this patch tested?

UT(TO add).

### CODE REVIEW

File: catalogs/hadoop-compatible-common/src/main/java/com/datastrato/gravitino/hadoop/authentication/AuthenticationConfig.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import com.datastrato.gravitino.config.ConfigEntry;
+import java.util.Map;
+
+public class AuthenticationConfig extends Config {
+
+  public static final String IMPERSONATION_ENABLE_KEY = "impersonation-enable";
+
+  public static final boolean DEFAULT_IMPERSONATION_ENABLE = false;
+
+  public static final ConfigEntry<Boolean> CHECK_INTERVAL_SEC_ENTRY =
```
Comment (qqqttt123) [inline catalogs/hadoop-compatible-common/src/main/java/com/datastrato/gravitino/hadoop/authentication/AuthenticationConfig.java:20]: I don't think we should share the configuration for Hive and Hadoop.

Comment (jerqi) [reply catalogs/hadoop-compatible-common/src/main/java/com/datastrato/gravitino/hadoop/authentication/AuthenticationConfig.java:20]: Actually they are different config options, they control different entities. They should be supported in different versions. Some config option may deprecated for the change of underlying systems. They are different.

Comment (yuqi1129) [reply catalogs/hadoop-compatible-common/src/main/java/com/datastrato/gravitino/hadoop/authentication/AuthenticationConfig.java:20]: Can someone else assist in leaving comments here? In all, I would like to share some common configuration keys between Hive and HDFS, maybe it's not very proper to make them all shareable. 

@FANNG1 @diqiu50

**Comment by jerryshao** [comment]
Is it ready for review?

**Comment by yuqi1129** [comment]
> Is it ready for review?

I'm afraid we need to add some tests using the HDFS cluster, not just a mini cluster here. If this does not matter, I think it's ready for review.

**Comment by jerryshao** [comment]
We could have a separate PR for e2e test, using mock test here to cover the logic should be enough.

File: catalogs/hadoop-compatible-common/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    exclude("org.apache.hadoop.hdfs.web")
+    exclude("javax.servlet", "servlet-api")
+    exclude("org.eclipse.jetty")
+  }
+  implementation(libs.hadoop3.client) {
+    exclude("com.sun.jersey")
+    exclude("org.apache.hadoop.hdfs.web")
+    exclude("javax.servlet", "servlet-api")
+    exclude("org.eclipse.jetty")
+  }
```
Comment (jerryshao) [inline catalogs/hadoop-compatible-common/build.gradle.kts:41]: I think Hive catalog is using hadoop2, what will be happened if you change to hadoop3 in this module and make hive catalog dependable on this module?

Comment (yuqi1129) [reply catalogs/hadoop-compatible-common/build.gradle.kts]: So, we need to implement a separate logic for Kerberos authentication in Hive and Hadoop catalogs?

File: catalogs/hadoop-compatible-common/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+plugins {
+  id("java")
+}
+
+group = "org.example"
+version = "0.6.0-SNAPSHOT"
```
Comment (jerryshao) [inline catalogs/hadoop-compatible-common/build.gradle.kts]: Why are these two lines required and the license header is 2023?

Besides, I think it is not so elegant to have a new module for code sharing, is there a better solution to do it?

Comment (yuqi1129) [reply catalogs/hadoop-compatible-common/build.gradle.kts]: Many JDBC-related catalogs share some common code and introduce a new catalog `catalog-jdbc-common`, I think we also do it. 

Besides, moving this module out of `catalogs` directory is not feasible. Currently, I haven't come up with a better one unless we will not share the code.

File: catalogs/hadoop-compatible-common/src/main/java/com/datastrato/gravitino/hadoop/authentication/AuthenticationConfig.java (hunk 1 above)
Comment (FANNG1) [reply catalogs/hadoop-compatible-common/src/main/java/com/datastrato/gravitino/hadoop/authentication/AuthenticationConfig.java]: could you provide more context?

File: catalogs/hadoop-compatible-common/build.gradle.kts (hunk 3 above)
Comment (qqqttt123) [reply catalogs/hadoop-compatible-common/build.gradle.kts]: Prefer not sharing code.

Comment (yuqi1129) [reply catalogs/hadoop-compatible-common/build.gradle.kts]: @jerryshao  What's your thought about code sharing here, do we need to share kerberos-related code between different catalog modules? thanks.

**Comment by yuqi1129** [comment]
The following point needs clarification:
- After introducing a common module that handles Kerberos authentication, the Hive catalog will depend on hadoop3(see below), is it acceptable? I think it's a bit risky to use a different release version for the Hive catalog.
image
- If the first option is not desirable, I believe we may not be able to share code between modules that use Kerberos authentication, can we accept duplicated code of Kerberos code logic here? 

@jerryshao 
I'm looking forward to your kind reply, thank you.

**Comment by jerryshao** [comment]
I would choose the option 2.

**Comment by yuqi1129** [comment]
@qqqttt123 @jerryshao 
The code has been updated, please help to review it if you have time, thanks.

**Comment by yuqi1129** [comment]
@jerryshao 
Please help to review this pr, The Hadoop e2e test depends on this PR.

**Comment by yuqi1129** [comment]
@jerryshao @qqqttt123 
Please help to review it, thanks.

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/KerberosConfig.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+          .doc("The fetch timeout of the kerberos connection")
+          .version(ConfigConstants.VERSION_0_3_0)
+          .intConf()
+          .checkValue(value -> value > 0, ConfigConstants.POSITIVE_NUMBER_ERROR_MSG)
+          .createWithDefault(2);
+
+  public static final ConfigEntry<Boolean> ENABLE_IMPERSONATION_ENTRY =
+      new ConfigBuilder(IMPERSONATION_ENABLE_KEY)
+          .doc("Whether to enable impersonation for the Hadoop catalog")
+          .version(ConfigConstants.VERSION_0_3_0)
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/KerberosConfig.java:67]: To fix the document, I will use another PR. @qqqttt123

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/KerberosConfig.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    return get(FETCH_TIMEOUT_SEC_ENTRY);
+  }
+
+  public static final Map<String, PropertyEntry<?>> KERBEROS_PROPERTY_ENTRIES =
+      new ImmutableMap.Builder<String, PropertyEntry<?>>()
+          .put(
+              IMPERSONATION_ENABLE_KEY,
+              PropertyEntry.booleanPropertyEntry(
+                  IMPERSONATION_ENABLE_KEY,
+                  "Enable user impersonation for Hive catalog",
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/KerberosConfig.java]: Fix

**Comment by yuqi1129** [comment]
All resolved, please help to double-check if it's okay now. 
@qqqttt123

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopProxyPlugin.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      throws Throwable {
+    try {
+      UserGroupInformation proxyUser;
+
+      if (UserGroupInformation.isSecurityEnabled() && ops != null) {
+
+        // The Gravitino server may use multiple KDC servers.
+        // The http authentication use one KDC server, the Hive catalog may use another KDC server.
+        // The KerberosAuthenticator will remove realm of principal.
+        // And then we add the realm of Hive catalog to the user.
```
Comment (jerqi) [inline catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopProxyPlugin.java:44]: The comment isn't right.

**Comment by jerqi** [comment]
You need to notice this https://github.com/apache/incubator-uniffle/pull/824

**Comment by yuqi1129** [comment]
> What configurations do you need if you support IAM? Could you configure it with kerberos at the same time? I also have concern the support of fileset Kerberos. We can't use keytabs of different KDC in the same catalog.

I would prefer to use another PR solution to support multiple KDCs and multiple tables in the Hadoop catalogs, I think the Hive catalog also needs to handle this issue.  

I may take some time to think about it and complete it.

**Comment by jerqi** [comment]
> > What configurations do you need if you support IAM? Could you configure it with kerberos at the same time? I also have concern the support of fileset Kerberos. We can't use keytabs of different KDC in the same catalog.
> 
> I would prefer to use another PR solution to support multiple KDCs and multiple tables in the Hadoop catalogs, I think the Hive catalog also needs to handle this issue.
> 
> I may take some time to think about it and complete it. Currently, we don't have enough time to handle it in this version.

Hive catalog doesn't need to handle this.

**Comment by yuqi1129** [comment]
> > > What configurations do you need if you support IAM? Could you configure it with kerberos at the same time? I also have concern the support of fileset Kerberos. We can't use keytabs of different KDC in the same catalog.
> > 
> > 
> > I would prefer to use another PR solution to support multiple KDCs and multiple key tables in the Hadoop catalogs, I think the Hive catalog also needs to handle this issue.
> > I may take some time to think about it and complete it. Currently, we don't have enough time to handle it in this version.
> 
> Hive catalog doesn't need to handle this.

I have added the following keys and other according changes:

```
public static final String AUTH_TYPE_KEY = "hadoop.authentication.type";
public static final String ENABLE_AUTH_KEY = "hadoop.authentication.enable";
```

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                  "Whether to enable authentication for Hadoop catalog",
+                  false,
+                  true,
+                  false,
+                  false,
+                  false))
+          .put(
+              AUTH_TYPE_KEY,
+              PropertyEntry.stringImmutablePropertyEntry(
+                  AUTH_TYPE_KEY, "The uri of key tab for the catalog", false, null, false, false))
```
Comment (jerqi) [inline catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java:62]: The comment isn't right.

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import com.datastrato.gravitino.Config;
+import com.datastrato.gravitino.config.ConfigBuilder;
+import com.datastrato.gravitino.config.ConfigConstants;
+import com.datastrato.gravitino.config.ConfigEntry;
+import com.datastrato.gravitino.connector.PropertyEntry;
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+
+public class AuthenticationConfig extends Config {
+  public static final String AUTH_TYPE_KEY = "hadoop.authentication.type";
```
Comment (jerqi) [inline catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java:17]: Maybe we should use `authentication.type`.

Comment (yuqi1129) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java:17]: This is the configuration of Hadoop catalog, why do we need to remove `hadoop` here?

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/KerberosConfig.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+import com.datastrato.gravitino.config.ConfigBuilder;
+import com.datastrato.gravitino.config.ConfigConstants;
+import com.datastrato.gravitino.config.ConfigEntry;
+import com.datastrato.gravitino.connector.PropertyEntry;
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+
+public class KerberosConfig extends AuthenticationConfig {
+  public static final String KEY_TAB_URI_KEY = "kerberos.keytab-uri";
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/KerberosConfig.java:17]: This makes sense.

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java (hunk 8 above)
Comment (jerqi) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java:17]: Hadoop catalog properties don't start with `hadoop`. You can refer to `location`. `hadoop` is rebudant information.

Comment (yuqi1129) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/kerberos/AuthenticationConfig.java:17]: Let me check it.

**Comment by jerqi** [comment]
LGTM, @jerryshao Could you check whether the config options are ok?

**Review by qqqttt123 — APPROVED** [review]

**Pull request merged**